cql3/query_backend.hh

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (C) 2023-present ScyllaDB
+ *
+ * Modified by ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: (AGPL-3.0-or-later and Apache-2.0)
+ */
+
+#pragma once
+
+#include "data_dictionary/data_dictionary.hh"
+
+namespace service {
+class forward_service;
+class migration_manager;
+class raft_group0_client;
+class storage_proxy;
+}
+
+namespace wasmtime {
+class Engine;
+}
+
+namespace wasm {
+class instance_cache;
+class alien_thread_runner;
+}
+
+namespace cql_transport::messages {
+class result_message;
+}
+
+namespace cql3 {
+
+using computed_function_values = std::unordered_map<uint8_t, bytes_opt>;
+
+// Abstract interface for the query backend, presented to the CQL3 code.
+//
+// It hides the details of how requests are executed.
+class query_backend {
+public:
+    class impl;
+
+private:
+    shared_ptr<impl> _impl;
+
+public:
+    explicit query_backend(shared_ptr<impl>);
+    ~query_backend();
+
+public:
+    data_dictionary::database db();
+    service::forward_service& forwarder();
+    service::migration_manager& get_migration_manager();
+    service::raft_group0_client& get_group0_client();
+    wasmtime::Engine& wasm_engine();
+    wasm::instance_cache& wasm_instance_cache();
+    wasm::alien_thread_runner& alien_runner();
+
+    //FIXME: get rid of this, by providing abstract methods for all proxy methods currently used directly.
+    service::storage_proxy& proxy();
```

Comment:
I think it's unrealistic. storage_proxy does quite a lot:
- walk over vnodes/tablets
- handle consistency level
- make decisions re range/point queries
- implement LWT

Reply:
@kbr-scylla tries it. He found out it would take a lot of code duplication.

Reply:
Tried what?

Reply:
We don't even need two separate implementations, just an instance that has the messaging_service pointers null. They'll never be called for local tables.

See _force_local below for a way to force it:

```c++
    future<rpc::tuple<foreign_ptr<lw_shared_ptr<reconcilable_result>>, cache_temperature>> make_mutation_data_request(lw_shared_ptr<query::read_command> cmd, gms::inet_address ep, clock_type::time_point timeout) {
        ++_proxy->get_stats().mutation_data_read_attempts.get_ep_stat(get_topology(), ep);
        if (_force_local || fbu::is_me(ep)) {
            tracing::trace(_trace_state, "read_mutation_data: querying locally");
            return _proxy->query_mutations_locally(_schema, cmd, _partition_range, timeout, _trace_state);
        } else {
            return _proxy->remote().send_read_mutation_data(netw::messaging_service::msg_addr{ep, 0}, timeout, _trace_state, *cmd, _partition_range);
        }
    }
```

Reply:
> We don't even need two separate implementations, just an instance that has the messaging_service pointers null. 
They'll never be called for local tables.

That sounds fine.

> 
> See _force_local below for a way to force it:
> 
> ```c++
>     future<rpc::tuple<foreign_ptr<lw_shared_ptr<reconcilable_result>>, cache_temperature>> make_mutation_data_request(lw_shared_ptr<query::read_command> cmd, gms::inet_address ep, clock_type::time_point timeout) {
>         ++_proxy->get_stats().mutation_data_read_attempts.get_ep_stat(get_topology(), ep);
>         if (_force_local || fbu::is_me(ep)) {
>             tracing::trace(_trace_state, "read_mutation_data: querying locally");
>             return _proxy->query_mutations_locally(_schema, cmd, _partition_range, timeout, _trace_state);
>         } else {
>             return _proxy->remote().send_read_mutation_data(netw::messaging_service::msg_addr{ep, 0}, timeout, _trace_state, *cmd, _partition_range);
>         }
>     }
> ```

Why would you force it instead of reporting an error?

Reply:
> > We don't even need two separate implementations, just an instance that has the messaging_service pointers null.
> > They'll never be called for local tables.
> 
> That sounds fine.

But then why do you need two instances? Have one that is created with ms set to null earlier and then, when ms + topology is initialized, set ms to the real value. All quires that are not local will fail while ms in null.

Reply:
Okay by me.

Though I think it's cleaner to have two instances, since a user-level query may involve a system-level query. storage_proxy is even now only semi-reentrant, since a later query may wait for an earlier query to complete. That's a deadlock scenario.

Reply:
I think we do need `_force_local`. You not only need to ensure messaging is not used (which is easy enough), but you need to ensure, topology is not used either at all. You need to ensure hints are not used and so on. Basically there is a long list of things that a local storage proxy is not allowed to even touch. The safest way to guarantee that is to have this `_force_local`. But then, if you have an `if` in each public method, you might as well have two classes at that point.

Reply:
The `query_backend` interface this PR introduces, allows a great deal of flexibility of how we approach this: we can have two instances of the same type, two instances of different types, or even a backend which completely bypasses storage proxy. We can even change our minds later. The CQL3 code doesn't care.

I think this interface has merits on its own. We can also move this to `service/` and make other frontends use it too (`Thrift`, `Alternator` and `Redis`).

Reply:
> Okay by me.
> 
> Though I think it's cleaner to have two instances, since a user-level query may involve a system-level query. 
storage_proxy is even now only semi-reentrant, since a later query may wait for an earlier query to complete. That's a deadlock scenario.

Do you mean back-pressure mechanism here? It does not limit parallelism. This is on the client. IIRC sp never waits for earlier query to complete before starting a new one. It may delay completion of a query as a back-pressure measure, but never local one since local one will have only one replica to wait for.

Reply:
> I think we do need `_force_local`. You not only need to ensure messaging is not used (which is easy enough), but you need to ensure, topology is not used either at all. You need to ensure hints are not used and so on. Basically there is a long list of things that a local storage proxy is not allowed to even touch. The safest way to guarantee that is to have this `_force_local`. But then, if you have an `if` in each public method, you might as well have two classes at that point.

The list is not long. As far as I see it contains may be three items: messaging service, feature service and topology (hints reply is an external service). And the later is more or less already the case. With topology over raft we do query through sp before topology is available. BTW messaging service also depends on the topology and sometimes we need to use it to figure out the topology. So are we going to have two instances of ms one to use before topology is knows and another is after?

And I do not understand `_force_local` thing. We should not force the wrong write to go local, we should return an error.

Reply:
> But then, if you have an if in each public method, you might as well have two classes at that point.

And duplicate quite a lot of pretty non trivial code? Or having them both call current code, but one will check that the query is local end error otherwise? The later will just add the indirection.

Reply:
The local variant of both reads and writes is pretty simple. All the complicated parts are in the coordinator side.

Reply:
It's easy to extract the local part to a function.